### DISCUSSION TITLE
TDD의 장점과 본인 경험을 들려주세요

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#37] [Test]
.

**Comment by jimi567** [comment]
이론적인 장점은 다음과 같습니다.
코드의 동작의 신뢰성을 빠르게 얻을 수 있습니다.
코드를 작성하기 전에 요구 사항을 보다 명확하게 파악할 수 있습니다.
TDD 의 코드는 유지보수성을 향상시킵니다. 각각의 기능에 대한 테스트 케이스가 항상 존재하기 때문에 코드 변경 시 기존 기능과 동일한 기능을 하는지 빠른 피드백을 받을 수 있습니다.
개발 생산성이 향상 됩니다. 초기에 테스트 작성에 들이는 시간이 생겨 오래 걸릴 수 있지만, 장기적으로 개발 시간을 단축시켜 주기 때문에 생산성을 크게 향상 시킵니다.

(아래는 제 경험입니다)
TDD로 개발하게 될 경우, 프로덕션 코드를 개발하는 단계에서는 확실히 시간이 오래 걸리긴 했지만, 리팩토링 과정에서 시간을 엄청 단축 시킬 수 있었습니다.
내가 리팩토링한 코드가 기존에 작성한 코드와 동일한 결과를 낸다는 확신을 빠르게 얻을 수 있었습니다. 또, 특정 케이스에서 원하는 예외가 발생하는 지를 체크하기에도 편리 했습니다. 
(사실 멋드러지게 유지보수 측면에서~ 협업 과정에서~ 이런 말 쓰고 싶은데 내가 경험이 없네,,)

**Comment by nayonsoso** [comment]
TDD를 하면 아래와 같은 장점이 있습니다.
1. 기능에 집중해서 구현하게 되므로 오버 엔지니어링을 방지할 수 있습니다.
2. 한번에 모든 것을 개발할 수 없으므로, TDD를 통해 단계를 나누어 개발의 순서를 제어할 수 있습니다.
3. 기능 명세서와 달리 실제로 돌아가는 코드에 대한 검증을 할 수 있다는 점에서 더 신뢰를 할 수 있습니다.

(아래는 제 경험입니다)
저는 장점 중에서 `기능에 집중해서 구현할 수 있다`와 `단계를 나눌 수 있다`는 점이 특히 와닿았어요.
미션을 할 때 구현을 하다가, 설계를 생각하고, 다시 구현을 하고, 설계를 생각해서 엎는 등...
구현과 설계를 왔다갔다 할 때가 있었습니다😓
그렇게 하다 보면 제가 지금 뭘 하고있었지~ 하고 방향을 잃곤 했습니다.

하지만 TDD를 적용하면서, 일단 테스트가 통과될 정도로 최소한의 코드를 작성하는 습관을 들였어요.
그렇다 보니 기능 자체에 초점을 맞출 수 있었고, 쓸데없는 생각들도 들지 않았습니다.
또한 테스트를 먼저 작성하니 내가 무엇을 테스트 할 것인지 명확해지는 느낌을 받았습니다.

**Comment by GIVEN53** [comment]
기능에 집중
리팩토링 시 부담 없음
코드에 대한 신뢰 상승
